### PR TITLE
Split `-Zchalk` flag into `-Ztrait-solver=(classic|chalk|next)` flag

### PR DESCRIPTION
**Pull request by compiler-errors**
We'll eventually need a way to select more than chalk + not-chalk. 

Does this need an MCP since it's touching a `-Z` flag? Or perhaps I should preserve `-Zchalk` for the time being... maybe I could make it a warning to use that flag? cc @rust-lang/types 

r? types

### CODE REVIEW

**Comment by compiler-errors** [comment]
Also could probably use advice on what to call the current solver. "stock" is perhaps not the best name, but couldn't think of anything better.

**Comment by jeffparsons** [comment]
> Also could probably use advice on what to call the current solver.

How about "classic", with an alias "default" that currently points to "classic" but that could be updated in future.

**Comment by jackh726** [comment]
"stock" -> "legacy"?

r=me either way

I don't think we need an MCP. I'm not sure if -Z options require them anyways. But regardless, this is definitely a types thing and pretty clearly falls under our plans, so this shouldn't come as a surprise to anyone.

**Comment by compiler-errors** [comment]
I actually prefer "classic" to "legacy".

I will probably also put up another PR which reintroduces `-Zchalk` as an opt that fails with an explicit message mentioning `-Ztrait-solver=chalk`, just in case people are confused -- I certainly would be. But that's a bit more invasive change, so I'd prefer if it's reviewed separately/we can discuss whether it's necessary.

@bors r=jackh726

File: compiler/rustc_session/src/options.rs [hunk 1]
```diff
@@ -880,6 +882,18 @@ mod parse {
         }
     }
 
+    pub(crate) fn parse_trait_solver(slot: &mut TraitSolver, v: Option<&str>) -> bool {
+        match v {
+            Some("stock") => *slot = TraitSolver::Classic,
```
Comment (klensy) [inline compiler/rustc_session/src/options.rs:887]: so, it's classic, not `stock`?

**Comment by klensy** [comment]
PR title should be fixed too?

File: compiler/rustc_session/src/options.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_session/src/options.rs:887]: :facepalm:

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 2]
```diff
@@ -1167,7 +1168,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     }
 
                     ty::PredicateKind::WellFormed(ty) => {
-                        if !self.tcx.sess.opts.unstable_opts.chalk {
+                        if self.tcx.sess.opts.unstable_opts.trait_solver != TraitSolver::Chalk {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1171]: the new solver can also error for `WellFormed` by themselves

**Pull request merged**